cranelift/docs/isle.md

```diff
@@ -0,0 +1,1306 @@
+    (C (D x)) -> (E x)
+    (C (F x)) -> (G x)
+```
+
+This set of rules will rewrite `(A (B (D 42)))` to `(C (D 42))`, then
+to `(E 42)` (via the first and third rules respectively).
+
+How is this useful? First, rewriting one term to another (here, `C` at
+the top level) that in turn appears in the left-hand side of other
+rules allows us to *factor* a "program" of term-rewriting rules in the
+same way that imperative programs are factored into separate
+functions.[^4] The usual advantages of a well-factored program, where
+each problem is solved with a small step that "reduces to a previously
+solved problem", apply here.
+
+Second, repeating the rewrite step is actually what grants
+term-rewriting its Turing-completeness: it allows for arbitrary
+control flow.[^5] This might be useful in cases where, for example, a
+term-rewriting program needs "loop" over a list of elements in the
+input: it can recurse and use intermediate terms to store state.
+
+While this full generality may not be used often in the
+domain-specific applications of term-rewriting that emphasize its
+pattern-matching (such as instruction selectors), the user should not
+be afraid to define and use intermediate terms -- rewriting into them,
+then defining additional rules to rewrite further -- when it helps to
+factor common behavior out of multiple rules, or aids in conceptual
+clarity.
+
+[^4]: In fact, ISLE actually compiles rules for different top-level
+      pattern terms (`(A ...)` and `(C ...)` in the example) into
+      separate Rust functions, so factoring rules to use intermediate
+      terms can provide code-size and compile-time benefits for the
+      ISLE-generated Rust code as well.
+      
+[^5]: The [lambda calculus' reduction
+      rules](https://en.wikipedia.org/wiki/Lambda_calculus#Reduction)
+      are a good example of this.
+
+### Application to Compilers: A Term is a Value; Rewrites are Lowerings
+
+So far this has been a fairly abstract introduction to term-rewriting
+systems as a general computing paradigm. How does this relate (or, how
+is it commonly mapped) to the instruction-selection problem?
+
+In a domain such as instruction selection, we manipulate terms that
+represent computations described by an IR, and the terms are
+eventually rewritten into terms that name specific machine
+instructions. We can think of each term as having a denotational value
+that that *is* that program value. Then, any rewrite is correct if it
+preserves the denotational value of the term.
+
+In other words, terms are just values, and rules specify alternative
+ways to compute the same values. We might have rewrite rules that
+correspond to common algebraic identities (`a + b` == `b + a`, and
+`a + 0` == `a`), for example. The main sort of rewrite rule, however,
+will be one that takes a machine-*independent* operator term and
+rewrites it into a machine-*dependent* instruction term. For example:
+
+```plain
+    (iadd a b) -> (isa.add_reg_reg a b)
```

Comment:
Can all ISLE terms/expressions operate on either CLIF or the machine dependent IR? That is, can we think of rules are operating on the language that is the union of the input and output languages? Or does something prevent compiler writers from adding a term that rewrites a machine dependent IR term back into a higher level term?

Reply:
ISLE itself is agnostic to the underlying IR, it just knows about extractors and constructors and various types that the user defines; so one could have `(decl compile_to_machine1 (CLIF) MachInst1)` and `(decl compile_to_machine2 (CLIF) MachInst2)` and `(decl optimize (CLIF) CLIF)` and define rules for each. At least with the first two, we have a type-level guarantee that lowering goes in the proper direction. The latter could have convergence issues but we could limit the number of passes rather than chase a fixpoint indefinitely. Later in the doc ("A Note on Heterogeneous Types") touches on this distinction from other TRS-based instruction selector systems; hopefully that clarifies a bit more!